File: vi/codes/swift/chapter_hashing/hash_map.swift
/**
 * File: hash_map.swift
 * Created Time: 2023-01-16
 * Author: nuomi1 (nuomi1@qq.com)
 */

import utils

@main
enum HashMap {
    /* Mã điều khiển */
    static func main() {
        /* Khởi tạo bảng băm */
        var map: [Int: String] = [:]

        /* Thao tác thêm */
        // Thêm cặp khóa-giá trị (key, value) vào bảng băm
        map[12836] = "Xiao Ha"
        map[15937] = "Xiao Luo"
        map[16750] = "Xiao Suan"
        map[13276] = "Xiao Fa"
        map[10583] = "Xiao Ya"
        print("\nAfter adding is complete, hash table is\nKey -> Value")
        PrintUtil.printHashMap(map: map)

        /* Thao tác truy vấn */
        // Đưa khóa vào bảng băm để lấy giá trị
        let name = map[15937]!
        print("\nInput student ID 15937, found name \(name)")

        /* Thao tác xóa */
        // Xóa cặp khóa-giá trị (key, value) khỏi bảng băm
        map.removeValue(forKey: 10583)
        print("\nAfter removing 10583, hash table is\nKey -> Value")
        PrintUtil.printHashMap(map: map)

        /* Duyệt bảng băm */
        print("\nTraverse key-value pairs Key->Value")
        for (key, value) in map {
            print("\(key) -> \(value)")
        }
        print("\nTraverse keys only Key")
        for key in map.keys {
            print(key)
        }
        print("\nTraverse values only Value")
        for value in map.values {
            print(value)
        }
    }
}

File: vi/codes/java/chapter_hashing/hash_map.java
/**
 * File: hash_map.java
 * Created Time: 2022-12-04
 * Author: krahets (krahets@163.com)
 */

package chapter_hashing;

import java.util.*;
import utils.*;

public class hash_map {
    public static void main(String[] args) {
        /* Khởi tạo bảng băm */
        Map<Integer, String> map = new HashMap<>();

        /* Thao tác thêm */
        // Thêm cặp khóa-giá trị (key, value) vào bảng băm
        map.put(12836, "Xiao Ha");
        map.put(15937, "Xiao Luo");
        map.put(16750, "Xiao Suan");
        map.put(13276, "Xiao Fa");
        map.put(10583, "Xiao Ya");
        System.out.println("\nAfter adding is complete, hash table is\nKey -> Value");
        PrintUtil.printHashMap(map);

        /* Thao tác truy vấn */
        // Nhập khóa vào bảng băm để lấy giá trị
        String name = map.get(15937);
        System.out.println("\nInput student ID 15937, query name " + name);

        /* Thao tác xóa */
        // Xóa cặp khóa-giá trị (key, value) khỏi bảng băm
        map.remove(10583);
        System.out.println("\nAfter removing 10583, hash table is\nKey -> Value");
        PrintUtil.printHashMap(map);

        /* Duyệt bảng băm */
        System.out.println("\nTraverse key-value pairs Key->Value");
        for (Map.Entry<Integer, String> kv : map.entrySet()) {
            System.out.println(kv.getKey() + " -> " + kv.getValue());
        }
        System.out.println("\nTraverse keys only Key");
        for (int key : map.keySet()) {
            System.out.println(key);
        }
        System.out.println("\nTraverse values only Value");
        for (String val : map.values()) {
            System.out.println(val);
        }
    }
}
